Show a problem indicator when the password length exceeds the maxLength.

shuklamaneesh23 opened the pull request:
This PR resolves the issue of displaying an error message when users try to set a password that exceeds the maximum allowed length. The implementation follows the approach suggested [here](https://github.com/zulip/zulip/pull/29408#issuecomment-2054198028).

### Summary of Changes

- **Previous Behavior**: Previously, if the password input exceeded the maximum length, it was silently truncated. This PR enhances the experience by increasing the `maxlength` of the password input field beyond the supported limit.
- **Improved User Feedback**: Now, instead of truncating input, the signup button is disabled when the password length exceeds the limit. The field continues to accept input, showing the remaining character count and enhancing user experience.
- **New Function `check_overflow_password`**: Introduced in `signup.ts`, this function monitors the password input length and displays the remaining character count when the length approaches or exceeds the maximum limit.
- **Real-Time Limit Feedback**: Uses `render_compose_limit_indicator` to provide real-time feedback on character limits.
- **Visual Indicator**: Added a new span with the ID `password-limit-indicator` to visually indicate the character limit status for the password input field.

Fixes: #27922.

Design Discussion : [CZO Link](https://chat.zulip.org/#narrow/channel/101-design/topic/problem.20indicator.20when.20the.20password.20length.20exceeds).

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
<details>
<summary>Register Form: </summary>



https://github.com/user-attachments/assets/4a3b8e8c-8e09-4ffa-a56c-141c5c7720c6



</details>
<details>
<summary>Password Reset Form: </summary>



https://github.com/user-attachments/assets/3f553e6f-c65b-4b54-9780-8523e2a4b196



</details>

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

shuklamaneesh23 commented:
Hey @sahil839 , could you please take a look at this draft PR and let me know if I am going into the right direction or not?

sahil839 commented:
I don't think this is the correct approach.

`MAX_PASSWORD_LENGTH` in `RegistrationForm` should not be changed and should be set to the actual maximum length that we allow. This is necessary to make sure that correct validation takes place at the API level.
Only the `max_length` parameter of the UI should be removed or set to a large value, and then the UI should be updated to be similar to what we have for compose box.

Regarding the backend part, we do not create user if the password passed in the request is larger than the allowed length, but the request returns 200. In fact, it returs 200 for all the invalid form cases. So, this might be something that we might change but I am not sure.

shuklamaneesh23 commented:
Thanks @sahil839 , will change the max password length for the front end only. 

Is the UI part as per expectation or need some change?

sahil839 commented on templates/zerver/reset_confirm.html:
```diff
@@ -40,19 +40,16 @@ <h1>{{ _('Set a new password.') }}</h1>
+                    <div class="progress" id="pw_strength">
+                        <div class="bar bar-danger" style="width: 10%;"></div>
                     </div>
+                    <span id="reset-password-limit-indicator" class="tippy-zulip-tooltip" data-tippy-content="{{ ('Maximum password length: 100 characters') }}"></span>
```
Don't think there is a need for two separate commits, all changes can be done in one commit, since these elements are eventually used in the second commit.

sahil839 commented on web/src/portico/signup.ts:
```diff
@@ -342,3 +343,102 @@ $(() => {
+
+function update_sign_up_button_status(): void {
+    $("#signup-button")
+        .toggleClass("disabled-signup-button", password_too_long)
```
I think there is no use of this class and the CSS can be added based on the `disabled` attribute itself.

sahil839 commented on web/src/portico/signup.ts:
```diff
@@ -342,3 +343,102 @@ $(() => {
+
+let password_too_long = false;
+
+function set_password_too_long(status: boolean, is_signup: boolean): void {
```
I think we can avoid some of these functions and just have the code directly in `check_overflow_password`.

sahil839 commented on web/src/portico/signup.ts:
```diff
@@ -342,3 +343,102 @@ $(() => {
     });
 });
+
+function get_or_set(
```
This function is also not needed for now. Can just use `.val().trim()` while checking whether the input field overflows. A couple of if-blocks are anyways not used.

shuklamaneesh23 replied:
I tried using this approach, but it only provided the static value of the input field. However, I need the dynamic values of the input. So, it didn’t work for me.  

That said, I have removed the unnecessary `if` blocks that were not being used.

sahil839 reviewed (commented):
Posted some comments. I am not sure about the current design and how the remaining character count is shown. Is that discussed somewhere? If not you should first get some feedback on it in #design on CZO.

shuklamaneesh23 commented:
Hello @sahil839,  

I have addressed all the comments mentioned above and have also attached the design thread from [CZO](https://chat.zulip.org/#narrow/channel/101-design/topic/problem.20indicator.20when.20the.20password.20length.20exceeds) in the PR description.  

Please let me know if there are any further improvements I can make.  
Thank you!

sahil839 commented on templates/zerver/register.html:
```diff
@@ -143,12 +143,14 @@ <h1>{{ _('Create your account') }}</h1>
                 </div>
                 {% elif password_required %}
                 <div class="input-box password-div">
+                    <div class="password-container">
```
`password-label-container` would be a better class name.

sahil839 commented on web/src/portico/signup.ts:
```diff
@@ -342,3 +343,71 @@ $(() => {
+    });
+}
+
+export function check_overflow_password(input_field: string, is_signup: boolean): number {
```
I tried and I see no problem in directly passing `$password_elem` to this function and then get the value using `val()`.

Also, I guess we can skip `is_signup` variable and directly use the input jquery element to find the indicator and button element using `.find()` and `.closest()` functions of jquery.

shuklamaneesh23 replied:
> I tried and I see no problem in directly passing `$password_elem` to this function and then get the value using `val()`.
> 

Thanks for this, it worked.

> Also, I guess we can skip `is_signup` variable and directly use the input jquery element to find the indicator and button element using `.find()` and `.closest()` functions of jquery.

I tried using this, but the code becomes messier compared to the current implementation since the DOM structure of the `register` and `reset-password` pages differs.

sahil839 replied:
I guess you can just add a common class to button like `submit-button` and then do something like `closest("form").find(".submit-button")` since a form should have only one button in most of the cases.

sahil839 commented on web/src/portico/signup.ts:
```diff
@@ -342,3 +343,71 @@ $(() => {
+    } else {
+        $indicator.text("");
+    }
+    return input_field.length;
```
Why is this being returned here?

sahil839 commented on web/styles/portico/portico_signin.css:
```diff
@@ -1461,3 +1461,34 @@ button#register_auth_button_gitlab {
+#password-limit-indicator {
+    &:not(:empty) {
+        font-size: 16px;
+        color: var(--color-limit-indicator);
```
Chrome browser console says that these colors are not defined. Maybe because the relevant file is not imported. You might need to copy those here.

sahil839 replied:
I am not sure about this, since previously we just copied the colors in `markdown.css`. But we can wait for Tim's feedback.

sahil839 commented on web/styles/portico/portico_signin.css:
```diff
@@ -1461,3 +1461,34 @@ button#register_auth_button_gitlab {
+        cursor: pointer;
+        height: 0;
+        margin-right: 5px;
+        z-index: 10;
```
Is z-index required here?

sahil839 commented on templates/zerver/reset_confirm.html:
```diff
@@ -28,10 +28,12 @@ <h1>{{ _('Set a new password.') }}</h1>
-                    <label for="id_new_password1" class="">{{ _('Password') }}</label>
+                    <div class="password-container">
+                        <span><label for="id_new_password1" class="">{{ _('Password') }}</label></span>
+                        <span id="reset-password-limit-indicator" class="tippy-zulip-tooltip" data-tippy-content="{{ ('Maximum password length: 100 characters') }}" data-tippy-placement="top"></span>
```
I am not sure what language we used for translation in portico pages, but you should atleast tag this for translation using `trans` tag like we do for other `data-tippy-content` attributes in this file.

sahil839 reviewed (commented):
Posted some comments.

shuklamaneesh23 commented:
Hey @sahil839 , could you please take another look? 
Thanks!

sahil839 commented on web/src/portico/signup.ts:
```diff
@@ -342,3 +343,54 @@ $(() => {
+
+export function check_overflow_password($password_elem: JQuery, is_signup: boolean): void {
+    const password = $password_elem.val();
+    if (typeof password !== "string") {
```
If you can change the type above to `JQuery<HTMLInputElement` then this might not be needed.

shuklamaneesh23 commented:
@sahil839, I have updated the code based on your suggestions.  
Thank you so much!

sahil839 commented on web/src/portico/signup.ts:
```diff
@@ -342,3 +343,49 @@ $(() => {
+}
+
+export function check_overflow_password($password_elem: JQuery<HTMLInputElement>): void {
+    const password = $password_elem.val() ?? "";
```
I guess `?? ""` is to handle typescript linter. We usually do `val()!` in existing code, as we do not expect `val()` to return null or undefined for any case.

sahil839 reviewed (commented):
Posted few minor comments.

I am not sure whether this will need any design improvements, but looks good otherwise.

shuklamaneesh23 commented:
Thanks @sahil839 for the suggestions.
I have updated those minor bugs.
Please let me know if any further modifications required!

sahil839 commented on templates/zerver/reset_confirm.html:
```diff
@@ -63,7 +65,7 @@ <h1>{{ _('Set a new password.') }}</h1>
                 <div class="input-box m-t-30">
                     <div class="centered-button">
-                        <button type="submit" class="" value="Submit">Submit</button>
+                        <button type="submit" class="register-button" value="Submit">Submit</button>
```
`submit-button` would be a better class here, so you can add the same class in `register.html` as well. Also, would good to use form IDs in CSS as well to avoid breaking any other UI as `submit-button` is somehwat generic class name.

sahil839 commented on templates/zerver/reset_confirm.html:
```diff
@@ -63,7 +65,7 @@ <h1>{{ _('Set a new password.') }}</h1>
                 <div class="input-box m-t-30">
                     <div class="centered-button">
-                        <button type="submit" class="" value="Submit">Submit</button>
+                        <button type="submit" class="register-button" value="Submit">Submit</button>
```
`register-button` does not seem the best class name here. Would be good to use some other name like `submit-button`, but then the CSS selector should have use some unique class or ID as `submit-button` class might be used in other places.

shuklamaneesh23 replied:
Hey @sahil839,  

I had previously done the same by assigning unique IDs to both buttons as `signup-button` and `reset-button` and using these ID selectors to define the CSS. However, you suggested [here](https://github.com/zulip/zulip/pull/32141#discussion_r1853777966) removing those new `id`s and using the existing `register-button` selector to define the disabled properties.  

Should I revert to the previous version where I defined new IDs for both buttons? 

Thanks!

sahil839 replied:
I was thinking of adding a common class to both buttons and use selector like `.password-reset .submit-button` and similarly for signup page as well, but using IDs is also fine.

shuklamaneesh23 commented:
Hello @sahil839 , updated the code to use the `ID`s , please let me know if any further modifications are required.
Thanks!

sahil839 commented:
Looks good. I am not sure if any design update is needed. I guess you can rebase the PR and then ask for design feedback from Alya.

shuklamaneesh23 commented:
Hey @alya, I need your feedback over the design of this feature.
Does it match with the design decision made in the [CZO](https://chat.zulip.org/#narrow/channel/101-design/topic/problem.20indicator.20when.20the.20password.20length.20exceeds).

Thank you!

alya commented:
Please post the latest design in that CZO thread for further discussion.
